File: src/collections_and_containers/cpp/heap/test/test_heap.cpp
#include "heap.h"
#include "gtest/gtest.h"

TEST(MaxHeapTest, EmptyHeap) {
  MaxHeap<int> heap;
  ASSERT_TRUE(heap.empty());
}

TEST(MaxHeapTest, AddingMultipleElements) {
  MaxHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);
  EXPECT_EQ(heap.size(), 3);

  EXPECT_EQ(heap.pop(), 3);
  EXPECT_EQ(heap.pop(), 2);
  EXPECT_EQ(heap.pop(), 1);

  ASSERT_TRUE(heap.empty());

  // adding the same elements in different order
  heap.push(3);
  heap.push(1);
  heap.push(2);

  EXPECT_EQ(heap.pop(), 3);
  EXPECT_EQ(heap.pop(), 2);
  EXPECT_EQ(heap.pop(), 1);
}

TEST(MaxHeapTest, PeekVsPop) {

  // show the difference between peek and pop
  MaxHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  EXPECT_EQ(heap.peek(), 3);
  EXPECT_EQ(heap.peek(), 3);
  EXPECT_EQ(heap.size(), 3);

  EXPECT_EQ(heap.pop(), 3);
  EXPECT_EQ(heap.pop(), 2);
  EXPECT_EQ(heap.pop(), 1);
  ASSERT_TRUE(heap.empty());
}

TEST(MaxHeapTest, PeekWhenEmpty) {
  MaxHeap<int> heap;
  ASSERT_THROW(heap.peek(), std::out_of_range);
}

TEST(MaxHeapTest, PopWhenEmpty) {
  MaxHeap<int> heap;
  ASSERT_THROW(heap.pop(), std::out_of_range);
}

TEST(MaxHeapTest, CopyConstructor) {
  MaxHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  MaxHeap<int> heap2(heap);

  EXPECT_EQ(heap2.pop(), 3);
  EXPECT_EQ(heap2.pop(), 2);
  EXPECT_EQ(heap2.pop(), 1);
}

TEST(MaxHeapTest, AssignmentOperator) {
  MaxHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  MaxHeap<int> heap2 = heap;

  EXPECT_EQ(heap2.pop(), 3);
  EXPECT_EQ(heap2.pop(), 2);
  EXPECT_EQ(heap2.pop(), 1);
}

TEST(MaxHeapTest, isMaxHeap) {
  MaxHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  EXPECT_TRUE(heap.isMaxHeap());
}

TEST(MinHeapTest, EmptyHeap) {
  MinHeap<int> heap;
  ASSERT_TRUE(heap.empty());
}

TEST(MinHeapTest, AddingMultipleElements) {
  MinHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  EXPECT_EQ(heap.size(), 3);

  EXPECT_EQ(heap.pop(), 1);
  EXPECT_EQ(heap.pop(), 2);
  EXPECT_EQ(heap.pop(), 3);

  ASSERT_TRUE(heap.empty());

  // adding the same elements in different order

  heap.push(3);
  heap.push(1);
  heap.push(2);

  EXPECT_EQ(heap.pop(), 1);
  EXPECT_EQ(heap.pop(), 2);
  EXPECT_EQ(heap.pop(), 3);

  ASSERT_TRUE(heap.empty());
}

TEST(MinHeapTest, PeekVsPop) {

  // show the difference between peek and pop
  MinHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  EXPECT_EQ(heap.peek(), 1);
  EXPECT_EQ(heap.peek(), 1);
  EXPECT_EQ(heap.size(), 3);

  EXPECT_EQ(heap.pop(), 1);
  EXPECT_EQ(heap.pop(), 2);
  EXPECT_EQ(heap.pop(), 3);
  ASSERT_TRUE(heap.empty());
}

TEST(MinHeapTest, PeekWhenEmpty) {
  MinHeap<int> heap;
  ASSERT_THROW(heap.peek(), std::out_of_range);
}

TEST(MinHeapTest, PopWhenEmpty) {
  MinHeap<int> heap;
  ASSERT_THROW(heap.pop(), std::out_of_range);
}

TEST(MinHeapTest, CopyConstructor) {
  MinHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  MinHeap<int> heap2(heap);

  EXPECT_EQ(heap2.pop(), 1);
  EXPECT_EQ(heap2.pop(), 2);
  EXPECT_EQ(heap2.pop(), 3);
}

TEST(MinHeapTest, AssignmentOperator) {
  MinHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  MinHeap<int> heap2 = heap;

  EXPECT_EQ(heap2.pop(), 1);
  EXPECT_EQ(heap2.pop(), 2);
  EXPECT_EQ(heap2.pop(), 3);
}

TEST(MinHeapTest, IsMinHeap) {
  MinHeap<int> heap;
  heap.push(1);
  heap.push(2);
  heap.push(3);

  ASSERT_TRUE(heap.isMinHeap());
}

File: src/collections_and_containers/python/heap/test/test_heap.py
from src.heap import Heap, MaxHeap, MinHeap
from unittest import TestCase
import unittest


class TestMaxHeap(unittest.TestCase):
    def test_empty_heap(self):
        heap = MaxHeap()
        self.assertTrue(heap.empty())

    def test_adding_multiple_elements(self):
        heap = MaxHeap()
        heap.push(1)
        heap.push(2)
        heap.push(3)
        self.assertEqual(heap.size(), 3)

        self.assertEqual(heap.pop(), 3)
        self.assertEqual(heap.pop(), 2)
        self.assertEqual(heap.pop(), 1)

        self.assertTrue(heap.empty())

        # adding the same elements in different order
        heap.push(3)
        heap.push(1)
        heap.push(2)

        self.assertEqual(heap.pop(), 3)
        self.assertEqual(heap.pop(), 2)
        self.assertEqual(heap.pop(), 1)

    def test_peek_vs_pop(self):
        # show the difference between peek and pop
        heap = MaxHeap()
        heap.push(1)
        heap.push(2)
        heap.push(3)

        self.assertEqual(heap.peek(), 3)
        self.assertEqual(heap.peek(), 3)
        self.assertEqual(heap.size(), 3)

        self.assertEqual(heap.pop(), 3)
        self.assertEqual(heap.pop(), 2)
        self.assertEqual(heap.pop(), 1)
        self.assertTrue(heap.empty())

    def test_peek_when_empty(self):
        heap = MaxHeap()
        self.assertRaises(IndexError, heap.peek)

    def test_pop_when_empty(self):
        heap = MaxHeap()
        self.assertRaises(IndexError, heap.pop)

    def test_is_max_heap(self):
        heap = MaxHeap()
        heap.push(1)
        heap.push(2)
        heap.push(3)

        self.assertTrue(heap.is_max_heap())


class TestMinHeap(unittest.TestCase):
    def test_empty_heap(self):
        heap = MinHeap()
        self.assertTrue(heap.empty())

    def test_adding_multiple_elements(self):
        heap = MinHeap()
        heap.push(1)
        heap.push(2)
        heap.push(3)
        self.assertEqual(heap.size(), 3)

        self.assertEqual(heap.pop(), 1)
        self.assertEqual(heap.pop(), 2)
        self.assertEqual(heap.pop(), 3)

        self.assertTrue(heap.empty())

        # adding the same elements in different order
        heap.push(3)
        heap.push(1)
        heap.push(2)

        self.assertEqual(heap.pop(), 1)
        self.assertEqual(heap.pop(), 2)
        self.assertEqual(heap.pop(), 3)

    def test_peek_vs_pop(self):
        # show the difference between peek and pop
        heap = MinHeap()
        heap.push(1)
        heap.push(2)
        heap.push(3)

        self.assertEqual(heap.peek(), 1)
        self.assertEqual(heap.peek(), 1)
        self.assertEqual(heap.size(), 3)

        self.assertEqual(heap.pop(), 1)
        self.assertEqual(heap.pop(), 2)
        self.assertEqual(heap.pop(), 3)
        self.assertTrue(heap.empty())

    def test_peek_when_empty(self):
        heap = MinHeap()
        self.assertRaises(IndexError, heap.peek)

    def test_pop_when_empty(self):
        heap = MinHeap()
        self.assertRaises(IndexError, heap.pop)

    def test_is_min_heap(self):
        heap = MinHeap()
        heap.push(1)
        heap.push(2)
        heap.push(3)

        self.assertTrue(heap.is_min_heap())


if __name__ == "__main__":
    unittest.main()
